Async compaction - Single Consolidated PR

bvaradar opened the pull request:
Please use it as reference to get a complete view of Async Compaction changes. I will refresh this PR whenever I make changes to individual PRs.

Contains all changes made for Async Compaction in single PR for reference.  

Contains following PRs:

1. FileSystemView and Timeline level changes : https://github.com/uber/hudi/pull/390
2. Use Avro version of  CompactionWorkload for storing them under .hoodie : https://github.com/bvaradar/hudi/pull/4
3. Use the pending workload from (2) to track fileIds with pending compaction : https://github.com/bvaradar/hudi/pull/5
4. Store Compaction workload from (2) in auxiliary folder named ".extra" under hoodie. Employ write-once mechanism to avoid race-conditions : https://github.com/bvaradar/hudi/pull/6
5. Cleaner and Archiver changes to recognize fileIds under compaction and avoid causing data-loss : https://github.com/bvaradar/hudi/pull/7
6. API support for performing Async Compaction : https://github.com/bvaradar/hudi/pull/8
7.  Compactor and MergeHandle changes to ensure compaction is performed on data-files as specified in the workload: https://github.com/bvaradar/hudi/pull/9
8. Hoodie CLI changes to inspect, schedule and execute compaction : https://github.com/bvaradar/hudi/pull/10

n3nash commented on hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java:
```diff
@@ -1069,4 +1102,152 @@ private HoodieTable getTableAndInitCtx() {
+      }
+      logger.info("Compacted successfully on commit " + compactionCommitTime);
+    } else {
+      logger.info("Compaction did not run for commit " + compactionCommitTime);
```
This log may be incorrect when autoCommit is turned to false.

n3nash replied:
@bvaradar

bvaradar commented:
@n3nash : I have addresses most of the review comments found in various Async compaction PRs. (Have replied to each comment to keep track of the comments addressed). I have not addressed comments around refactoring HoodieWriteClient/HoodieCleanHelper alone( there is an issue opened for this).  Please review and let @vinothchandar and myself know about your comments.

n3nash commented:
@bvaradar Just went over the comments, did not do a full pass at it again. Left a couple of comments that need some discussion rest LGTM after which @vinothchandar can merge it!

n3nash commented:
@bvaradar @vinothchandar Looks good to be merged, finally!

vinothchandar commented:
Awesome, done!!
